### PR TITLE
Remove WPMediaPicker dependency from Tenor picker

### PR DESCRIPTION
**Pull request by kean**
This PR reimplements Tenor (Free GIFs) picker without the `WPMediaPicker` framework.

I took some liberties with the design and tweaked it a little bit to match the recently redesigned screens.

https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/a90af3c5-c064-4959-b235-21fb4ee2eaf7

## Changes

- Add `ExternalMediaPickerViewController` as a replacement for `TenorPicker` based on the `WPMediaPicker` framework
- Add `MediaPreviewController` modeled after `QLPreviewController`. I’m using `WPImageViewController` to display images, which is part of the app (not WPMediaPicker), so it won’t prevent us from removing it.
- Integrate the new picker in the following flows: Aztec and Gutenberg post editor, Site Media screen
- Update MediaImageService to support images from external sources. It takes advantage of all the improvements
- Extract `ImageDecoder` from `MediaImageService`
- Improve `ImageDownloader` by moving its work to the background, adding image decompression in the background
- Add `ImageView` to managing image downloads using the new architecture and design: solid gray placeholders, etc

## Known Issues

I think pagination in Tenor is not working, but it's a production issue. I'm not prioritizing fixing it because this screen is not used often.

## Testing

For every of the following flows:

- Add Image to Gutenberg
- Add Gallery to Gutenberg
- Add image to Aztec
- Add images to Site Media

Test the following:

- Verify that the new picker is displayed
- Verify that "Cancel" dismisses the screen and doesn't add any assets if the selection is made
- Verify that for multiple selection, the selection it oredered
- Verify that for multiple selection, a toolbar with a number of selected of items is displayed
- Verify that when you tap a toolbar button indicating the selection count the fullscreen gallery is displayed
- Verify that the fullscreen gallery renders GIFs with animation
- Verify that you can zoom in in the gallery
- Verify that when you tap "Done" the selected image are uploaded

## Regression Notes
1. Potential unintended areas of impact: Tenor (Aztec and Gutenberg post editor, Site Media screen)
2. What I did to test those areas of impact (or what existing automated tests I relied on): manual
3. What automated tests I added (or what prevented me from doing so): n/a

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

**Comment by kean** [comment]
I apologize for the 1000+ sloc PR, but it was the best way to test it in one go. I don't want to overburden anyone with testing.

**Review by momo-ozawa — COMMENTED** [review]
Hey @kean, the testing steps work as described, but I noticed the media screen doesn't animate the gifs you uploaded. Also notices are displayed a bit too high since we hide the bottom tabs on the media screen

https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/4046e477-8040-4ae8-bda5-02ed342a87f2

**Comment by kean** [comment]
>  doesn't animate the gifs you uploade

This is an existing issue in 23.7; I opened it here https://github.com/wordpress-mobile/WordPress-iOS/issues/22076. 

The Site Media screen now displays GIFs with animation an no "GIF" badge – to match wp.com – and all the thumbnails returned by the backend are animatable GIFs. But the local code for generating the thumbnails creates static images as thumbnails. I will address it separately as it's not introduce in this PR.

File: WordPress/Classes/Utility/Media/ImageDownloader.swift [hunk 1]
```diff
@@ -1,111 +1,145 @@
...
+    /// Returns an image from the memory cache.
     ///
-    private let session = URLSession(configuration: .default)
+    /// - note: Use it to retrieve the image synchronously, which is no not possible
+    /// with the async functions.
+    nonisolated func cachedImage(for imageURL: URL, size: CGSize? = nil) -> UIImage? {
+        cache[makeKey(for: imageURL, size: size)]
+    }
 
+    nonisolated func setCachedImage(_ image: UIImage?, for imageURL: URL, size: CGSize? = nil) {
```
Comment (momo-ozawa) [inline WordPress/Classes/Utility/Media/ImageDownloader.swift:91]: Is it safe to mark this setter as nonisolated?

File: WordPress/Classes/Utility/Media/ImageDownloader.swift [hunk 2]
```diff
@@ -136,8 +170,32 @@ class AnimatedImageWrapper: UIImage {
...
+// MARK: - Helpers
+
+protocol ImageDownloaderTask {
+    func cancel()
+}
+
+extension Operation: ImageDownloaderTask {}
+extension URLSessionTask: ImageDownloaderTask {}
+
+private struct AnonumousImageDownloadTask: ImageDownloaderTask {
```
Comment (momo-ozawa) [inline WordPress/Classes/Utility/Media/ImageDownloader.swift]:
```suggestion
private struct AnonymousImageDownloadTask: ImageDownloaderTask {
```

File: WordPress/Classes/Utility/Media/ImageDownloader.swift (hunk 1 above)
Comment (kean) [reply WordPress/Classes/Utility/Media/ImageDownloader.swift:91]: `MemoryCache` and the underlying `NSCache` are thread-safe, so it should be, yes.

File: WordPress/Classes/Utility/Media/ImageDownloader.swift (hunk 2 above)
Comment (kean) [reply WordPress/Classes/Utility/Media/ImageDownloader.swift]: Fixed & I fixed GIF thumbnails here https://github.com/wordpress-mobile/WordPress-iOS/pull/22083

**Review by momo-ozawa — APPROVED** [review]

**Pull request merged**